Rework crawl page migration

tw4l opened the pull request:
Fixes #2406 

Converts migration 0042 to launch a background job (parallelized across several pods) to migrate all crawls by optimizing their pages and setting `version: 2` on the crawl when complete.

Also Optimizes MongoDB queries for better performance.

Migration Improvements:

- Add `isMigrating` and `version` fields to `BaseCrawl`
- Add new background job type to use in migration with accompanying `migration_job.yaml` template that allows for parallelization
- Add new API endpoint to launch this crawl migration job, and ensure that we have list and retry endpoints for superusers that work with background jobs that aren't tied to a specific org
- Rework background job models and methods now that not all background jobs are tied to a single org
- Ensure new crawls and uploads have `version` set to `2`
- Modify crawl and collection replay.json endpoints to only include fields for replay optimization (`initialPages`, `pageQueryUrl`, `preloadResources`) if all relevant crawls/uploads  have `version` set to `2`
- Remove `distinct` calls from migration pathways
- Consolidate collection recompute stats

Query Optimizations:
- Remove all uses of $group and $facet
- Optimize /replay.json endpoints to precompute preload_resources, avoid fetching crawl list twice
- Optimize /collections endpoint by not fetching resources 
- Rename /urls -> /pageUrlCounts and avoid $group, instead sort with index, either by seed + ts or by url to get top matches.
- Use $gte instead of $regex to get prefix matches on URL
- Use $text instead of $regex to get text search on title
- Remove total from /pages and /pageUrlCounts queries by not using $facet


Remaining work:

- [x] Testing
- [x] Add documentation on migration, how to check if it succeeds, how to restart if necessary (**this should also be linked to from the release notes!)**

ikreymer committed:
migration fixes:
- use group instead of distinct in unique page stats, page filename migration
- consolidate collection recompute stats into single function
- bump to 1.14.0-beta.2

ikreymer committed:
bump version to 1.14.0-beta.2

ikreymer committed:
bump to 1.14.0-beta.3
add timeout for backend gunicorn worker

update

ikreymer committed:
typo fix

ikreymer committed:
empty filenames migration: use existing bulk re_add method for faster migration

ikreymer committed:
bump to 1.14.0-beta.4

tw4l committed:
Add version to BaseCrawl, set to 2 for new crawls/uploads

Also adds isMigrating, which we'll use in migration 0042

tw4l committed:
Use optimize pages background job in 0042

- Adds new optimize pages background job that updates crawl pages
and sets version on updated crawls to 2
- New job uses a new migration_job.yaml template with parallelism
set to 3
- Update BackgroundJob model and some ops methods to allow for
creating and retrying a background job with no oid
- Add new API endpoint to retry one specific background job that
isn't tied to a specific org (superuser-only)

tw4l committed:
Hacky workaroud: Pass default org to retry_background_job

Now that we have background jobs that aren't tied to a specific org,
much of the background_jobs module needs to be reworked to account
for that.

That will take some time, so for now, so that we can test the
migration, we just pass the default org to retry_background_job
if the job doesn't have an oid set.

tw4l committed:
Make sure job failure emails are sent even if no oid

tw4l committed:
Fix typing error

tw4l committed:
Only include replay optimization fields if all pages are optimized

Only include initialPages, pagesQueryUrl, and preloadResources in
replay.json responses for crawls and collections if all of the
relevant crawls have version set to 2.

tw4l committed:
Set isMigrating at same time as pulling next crawl

tw4l committed:
Redo background_jobs typing, add superuser non-org list endpoint

tw4l committed:
Remove crawl_type from optimize_pages bg job

tw4l committed:
Handle running crawls in page optimization

tw4l committed:
Remove isMigrating filter from running crawl check

tw4l committed:
Fix bg job retry

tw4l committed:
Cast job to right type in retry

tw4l committed:
Add API endpoint to launch migrate crawls job

tw4l committed:
Use SuccessResponseId model to include job id in response

tw4l committed:
Add first draft of upgrade notes in docs

ikreymer committed:
cleanup:
- simplify group query for unique pages
- remove unused code
- convert array index to getter

ikreymer committed:
add logging to optimize pages job

ikreymer committed:
logging

ikreymer committed:
check for empty result

ikreymer committed:
version: bump to 1.14.0-beta.5

ikreymer committed:
fix response model

ikreymer committed:
add retries for add_crawl_pages_to_db_from_wacz()
add typing

ikreymer committed:
back to single background_job.yaml, just pass scale as param

ikreymer committed:
fix parallelism setting

ikreymer committed:
retry at the wacz level, not crawl

ikreymer committed:
bg jobs: make single crawl readd a background job as well, if crawl_id is provided,
process single crawl

ikreymer committed:
add sort

ikreymer committed:
better call

ikreymer committed:
logging

ikreymer committed:
ensure pages are streamed from each wacz on demand

ikreymer committed:
ignore dupes on insertMany

ikreymer committed:
attempt dedup within wacz

ikreymer committed:
ignore dupe errors?

ikreymer committed:
update logging, add crawl_id + url index

ikreymer committed:
lint fix

ikreymer committed:
optimize pages: don't use $facet, don't compute actual total

ikreymer committed:
add index for default sort

tw4l commented on backend/btrixcloud/crawlmanager.py:
```diff
@@ -163,25 +162,45 @@ async def run_re_add_org_pages_job(
+    async def run_optimize_pages_job(
+        self, existing_job_id: Optional[str] = None, scale=3
+    ) -> str:
+        """run job to recalculate storage stats for the org"""
```
```suggestion
        """run job to optimize crawl pages"""
```

tw4l commented on backend/btrixcloud/pages.py:
```diff
@@ -104,7 +114,13 @@ async def add_crawl_pages_to_db_from_wacz(self, crawl_id: str, batch_size=100):
+                        await self._add_pages_to_db(
+                            crawl_id, pages_buffer, ordered=False
+                        )
+                    # pylint: disable=broad-exception-caught
+                    except Exception as e:
+                        print("Error inserting, probably dupe", e)
```
This might be a bit overly broad, especially since we're clearing the buffer even if there's an exception. Is there a chance the batch might not be written to the database that we should be handling here?

tw4l commented on backend/btrixcloud/pages.py:
```diff
@@ -113,65 +129,24 @@ async def add_crawl_pages_to_db_from_wacz(self, crawl_id: str, batch_size=100):
+                try:
+                    await self._add_pages_to_db(crawl_id, pages_buffer, ordered=False)
+                # pylint: disable=broad-exception-caught
+                except Exception as e:
+                    print("Error inserting, probably dupe", e)
```
Same as above. Is there a specific exception for duplicate records we can catch, or results of the query we can inspect?

tw4l replied:
If entire batches of pages are regularly dupes, there might be something off...

tw4l commented on backend/btrixcloud/pages.py:
```diff
@@ -757,34 +733,40 @@ async def list_collection_pages(
+        # except (IndexError, ValueError):
+        #    total = 0
+
+        total = await self.pages.estimated_document_count()
```
Hm, I know this is faster, but won't it return the estimated total count of all pages, not an estimated total count of all the pages available that match the query? Not sure it's useful to include at all at that point for pagination purposes.

tw4l replied:
If we don't need an accurate total for this endpoint (which I think is technically true for now, though will make general API usage harder with pagination) it'd be better to remove it entirely from the response and model than to give a wrong number like what would be provided by `self.pages.estimated_document_count()`.

ikreymer replied:
removed, not including totals for this reason.

ikreymer committed:
remove uniquge page counts from org metrics

ikreymer committed:
update model

ikreymer committed:
remove resources from collection list

ikreymer committed:
optimize coll /replay.json
- move preloadResources to be precomputed in update_collection_counts_and_tags()
- only query list of crawls once, reuse ids
- remove facet from list_collection_pages(), support passing in crawlIds

ikreymer committed:
optimize page snapshot query:
- rename /urls -> /pageSnapshots
- remove facet, just return results
- frontend: update to new model, don't check total, just look at results list length

ikreymer committed:
crawl /replay.json optimization:
- rename list_collection_pages -> list_replay_query_pages to be used for replay page querying for collection or single crawl, no page totals
- use list_replay_query_pages in crawl /replay.json

tw4l commented on backend/btrixcloud/colls.py:
```diff
@@ -797,13 +797,13 @@ async def list_urls_in_collection(
         crawl_ids = await self.get_collection_crawl_ids(coll_id)
 
         match_query: dict[str, object] = {"oid": oid, "crawl_id": {"$in": crawl_ids}}
-        sort_query: dict[str, int] = {"count": -1, "_id": 1}
```
This sorting was a fix by Sua to make the thumbnail/home page selection sort order reasonable, I don't think we want to change it here as it'll negatively impact the frontend.

ikreymer committed:
cleanup

ikreymer committed:
fix typo

ikreymer committed:
lint

emma-sg committed:
Delay rendering components doing data fetching in collection settings dialog until open (#2414)

Delays rendering contents of the collection settings dialog until
actually needed. Various fetches that internal components were running
were causing slowdowns in other parts of the app, so this should resolve
some of that.

ikreymer committed:
optimize page snapshots, add non-group alternative

ikreymer committed:
fix page_count update?

ikreymer committed:
tweak page snapshot sort order

ikreymer committed:
use gte

ikreymer committed:
readd unquote

ikreymer committed:
lint fix

ikreymer committed:
tweak

ikreymer committed:
more cleanup:
- restore pages response models to include 'items' again
- rename endpoint to pageSnapshots -> pageUrlCounts
- fix tests to not check 'total'

ikreymer committed:
test fix

ikreymer committed:
test fix

ikreymer committed:
Update backend/btrixcloud/crawlmanager.py

Co-authored-by: Tessa Walsh <tessa@bitarchivist.net>

ikreymer committed:
Pages Query: use $text search or prefix search (#2415)

Instead of regex search on url and title:
- If search string starts with https:// or http://, do a prefix search
on URL
- Otherwise, do a $text search on title, sort by match score.

---------

Co-authored-by: Tessa Walsh <tessa@bitarchivist.net>

tw4l committed:
Fix typo in response

tw4l committed:
Add test asserts for crawl/upload version field

ikreymer committed:
remove dupe index

ikreymer commented on backend/btrixcloud/crawlmanager.py:
```diff
@@ -163,25 +163,45 @@ async def run_re_add_org_pages_job(
+        )
+
+    async def run_optimize_pages_job(
+        self, existing_job_id: Optional[str] = None, scale=3
```
Should we default this to 1, but allow overriding via helm chart? For default deployment.

ikreymer committed:
set pages migration job to run at scale 1 by default
add migration_jobs_scale to customize this setting
chart values: document rerun_from_migration and migration_jobs_scale
chart: remove unneeded background worker timeout customization

ikreymer reviewed (approved):
🚀
